crates/turbopack-ecmascript/src/tree_shake/mod.rs

```diff
@@ -0,0 +1,719 @@
+
+    match &*parse_result {
+        ParseResult::Ok { program, .. } => {
+            if let Program::Module(module) = program {
+                let (mut dep_graph, items) = Analyzer::analyze(module);
+
+                dep_graph.handle_weak(true);
+
+                let (data, deps, modules) =
+                    dep_graph.split_module(&format!("./{filename}").into(), &items);
+
+                return Ok(SplitResult {
+                    data,
+                    deps,
+                    modules,
+                    parsed,
+                }
+                .cell());
+            }
+        }
+        _ => {}
+    }
+
+    Ok(SplitResult {
+        data: Default::default(),
+        deps: Default::default(),
+        modules: Default::default(),
+        parsed,
+    }
+    .cell())
+}
+
+#[turbo_tasks::function]
+pub(super) async fn part_of_module(
+    split_data: SplitResultVc,
+    chunk_id: Option<u32>,
+) -> Result<ParseResultVc> {
+    let split_data = split_data.await?;
+
+    let chunk_id = match chunk_id {
+        Some(v) => v,
+        None => return Ok(split_data.parsed),
+    };
+
+    let parsed = split_data.parsed.await?;
+
+    match &*parsed {
+        ParseResult::Ok {
+            globals,
+            comments,
+            source_map,
+            eval_context,
+            ..
+        } => {
+            // Not a module
+            if split_data.modules.is_empty() {
+                return Ok(split_data.parsed);
+            }
+
+            let program = Program::Module(split_data.modules[chunk_id as usize].clone());
+            let eval_context = EvalContext::new(&program, eval_context.unresolved_mark);
+
+            Ok(ParseResultVc::cell(ParseResult::Ok {
+                program,
+                globals: globals.clone(),
+                comments: comments.clone(),
+                source_map: source_map.clone(),
+                eval_context,
+            }))
+        }
+        ParseResult::Unparseable => Ok(ParseResultVc::cell(ParseResult::Unparseable)),
+        ParseResult::NotFound => Ok(ParseResultVc::cell(ParseResult::NotFound)),
+    }
+}
+
+impl EcmascriptModulePartAssetVc {
+    pub async fn from_splitted(
+        module: EcmascriptModuleAssetVc,
+        part: ModulePartVc,
+    ) -> Result<Self> {
+        let split_data = split(module.origin_path(), module.parse());
+        let result = split_data.await?;
+        let part = part.await?;
+
+        let key = match &*part {
+            ModulePart::ModuleEvaluation => Key::ModuleEvaluation,
+            ModulePart::Export(export) => Key::Export(export.await?.to_string()),
+        };
+
+        let chunk_id = match result.data.get(&key) {
+            Some(id) => *id,
+            None => return Err(anyhow!("could not find chunk id for module part {:?}", key)),
+        };
+
+        Ok(EcmascriptModulePartAsset {
+            full_module: module,
+            chunk_id,
+            split_data,
+        }
+        .cell())
```

Comment:
The `split` call could be deferred by storing `ModulePartVc` instead of the `part_id`...

`split` can be called in `content()` and `references()`.

I want to have the constructor as cheap as possible... We want to be lazy...